### PR TITLE
Add shellcheck to ci

### PR DESCRIPTION
**Pull request by phlax**
Add shellcheck to ci and fix some scripts that fail linting

This PR adds the framework for running shellcheck in ci

Im not sure if where i have hooked it in is the best place, apart from anything shellcheck does not give very reliable fixes - so it does not provide a patch to remedy as other code formatting ci does.

I have fixed some of the low-hanging fruit, but figured that it would probably be a good idea to fix other parts in separate PRs as they touch other parts of the code base

Signed-off-by: Ryan Northey <ryan@synca.io>

Commit Message: Add shellcheck to ci
Additional Description:
Risk Level: low/medium
Testing: 
Docs Changes:
Release Notes:
[Optional Runtime guard:]
[Optional Fixes #Issue] Partial fix for #7793 
[Optional Deprecated:]

### CODE REVIEW

File: ci/run_envoy_docker.sh [hunk 1]
```diff
@@ -30,5 +30,10 @@ docker run --rm ${ENVOY_DOCKER_OPTIONS} -e HTTP_PROXY=${http_proxy} -e HTTPS_PRO
   -e ENVOY_SRCDIR -e ENVOY_BUILD_TARGET -e SYSTEM_PULLREQUEST_TARGETBRANCH -e SYSTEM_PULLREQUEST_PULLREQUESTNUMBER \
   -e GCS_ARTIFACT_BUCKET -e BUILD_SOURCEBRANCHNAME -e BAZELISK_BASE_URL -e ENVOY_BUILD_ARCH -e SLACK_TOKEN -e BUILD_URI\
   -e REPO_URI -v "$PWD":/source --cap-add SYS_PTRACE --cap-add NET_RAW --cap-add NET_ADMIN "${ENVOY_BUILD_IMAGE}" \
-  /bin/bash -lc "groupadd --gid $(id -g) -f envoygroup && useradd -o --uid $(id -u) --gid $(id -g) --no-create-home \
-  --home-dir /build envoybuild && usermod -a -G pcap envoybuild && sudo -EHs -u envoybuild bash -c \"cd /source && $*\""
+  /bin/bash -lc "\
+  	    apt-get update \
+	    && apt-get install --no-install-recommends -qq -y shellcheck \
```
Comment (lizan) [inline ci/run_envoy_docker.sh:35]: Don't apt-get install here, can you open a PR to envoy-build-tools?

File: ci/do_ci.sh [hunk 2]
```diff
@@ -367,6 +367,7 @@ elif [[ "$CI_TARGET" == "check_format" ]]; then
   echo "check_format_test..."
   ./tools/code_format/check_format_test_helper.sh --log=WARN
   echo "check_format..."
+  ./tools/code_format/check_shellcheck_format.sh
```
Comment (lizan) [inline ci/do_ci.sh:370]: Can it do some auto fix?

**Review by lizan — COMMENTED** [review]
Nice, thanks for doing this.

File: ci/do_ci.sh (hunk 2 above)
Comment (phlax) [reply ci/do_ci.sh:370]: it has a `-f diff` flag, but i dont think you would want it to autofix.

Shellcheck can often get false positives and the suggestions will often not work the same as the code ime.

i can add the diff flag at least - i had originally not intended to output the shellcheck output, just the result - but some things were failing differently on azure due to different shellcheck version, so i added it back.

ill update now with diff, and some small cleanup...

File: ci/run_envoy_docker.sh (hunk 1 above)
Comment (phlax) [reply ci/run_envoy_docker.sh:35]: yep...

File: ci/do_ci.sh (hunk 2 above)
Comment (phlax) [reply ci/do_ci.sh:370]: hmm, it turns out the version of shellcheck in bionic doesnt have diff:

https://dev.azure.com/cncf/envoy/_build/results?buildId=49942&view=logs&j=9fd74226-df5b-5f29-39da-76fc341cc763&t=068b98b1-acf0-52d0-fc6e-f6d1a5b1d5ef&l=113

https://dev.azure.com/cncf/envoy/_build/results?buildId=49942&view=logs&j=9fd74226-df5b-5f29-39da-76fc341cc763&t=068b98b1-acf0-52d0-fc6e-f6d1a5b1d5ef&l=132

https://github.com/koalaman/shellcheck/issues/1220#issuecomment-615030110

Comment (lizan) [reply ci/do_ci.sh:370]: Ah ok, then lets leave this as it is, can you add a TODO, so we can have that when we update to newer version of shellcheck?

File: ci/run_envoy_docker.sh (hunk 1 above)
Comment (lizan) [reply ci/run_envoy_docker.sh:35]: build-tools PR merged.

File: ci/do_ci.sh (hunk 2 above)
Comment (phlax) [reply ci/do_ci.sh:370]: yep

**Comment by phlax** [comment]
wierdly, shellcheck is failing a script that is only present in master, and not in this PR

ive disabled it for now, but im wondering why the script appeared and was being checked.

**Comment by lizan** [comment]
> wierdly, shellcheck is failing a script that is only present in master, and not in this PR
> 
> ive disabled it for now, but im wondering why the script appeared and was being checked.

This is expected, Azure pipelines (or GitHub Actions) runs CI on PR merge commit instead of PR head.

**Review by lizan — APPROVED** [review]

**Pull request merged**